scipy/stats/_distribution_infrastructure.py

```diff
@@ -0,0 +1,5240 @@
+        squeezed_base_shape = max.shape
+
+        if region == 'typical':
+            typical = getattr(self, 'typical')
+            a, b = typical.get_numerical_endpoints(parameter_values)
+            a, b = np.broadcast_arrays(a, b)
+            min_here = np.asarray(a.squeeze())
+            max_here = np.asarray(b.squeeze())
+            z_in = typical.draw(n_in, 'in', min_here, max_here, squeezed_base_shape,
+                                rng=rng)
+        else:
+            z_in = domain.draw(n_in, 'in', min, max, squeezed_base_shape, rng=rng)
+        z_on = domain.draw(n_on, 'on', min, max, squeezed_base_shape, rng=rng)
+        z_out = domain.draw(n_out, 'out', min, max, squeezed_base_shape, rng=rng)
+        z_nan= domain.draw(n_nan, 'nan', min, max, squeezed_base_shape, rng=rng)
+
+        z = np.concatenate((z_in, z_on, z_out, z_nan), axis=0)
+        z = rng.permuted(z, axis=0)
+
+        z = np.reshape(z, tuple(shape_expansion) + squeezed_base_shape)
+        z = np.moveaxis(z, new_base_singletons, base_singletons)
+        return z
+
+    @abstractmethod
+    def validate(self, arr):
+        raise NotImplementedError()
+
+
+class _RealParameter(_Parameter):
+    r""" Represents a real-valued parameter.
+
+    Implements the remaining methods of _Parameter for real parameters.
+    All attributes are inherited.
+
+    """
+    def validate(self, arr, parameter_values):
+        r""" Input validation/standardization of numerical values of a parameter.
+
+        Checks whether elements of the argument `arr` are reals, ensuring that
+        the dtype reflects this. Also produces a logical array that indicates
+        which elements meet the requirements.
+
+        Parameters
+        ----------
+        arr : ndarray
+            The argument array to be validated and standardized.
+        parameter_values : dict
+            Map of parameter names to parameter value arrays.
+
+        Returns
+        -------
+        arr : ndarray
+            The argument array that has been validated and standardized
+            (converted to an appropriate dtype, if necessary).
+        dtype : NumPy dtype
+            The appropriate floating point dtype of the parameter.
+        valid : boolean ndarray
+            Logical array indicating which elements are valid (True) and
+            which are not (False). The arrays of all distribution parameters
+            will be broadcasted, and elements for which any parameter value
+            does not meet the requirements will be replaced with NaN.
+
+        """
+        arr = np.asarray(arr)
+
+        valid_dtype = None
+        # minor optimization - fast track the most common types to avoid
+        # overhead of np.issubdtype. Checking for `in {...}` doesn't work : /
+        if arr.dtype == np.float64 or arr.dtype == np.float32:
+            pass
+        elif arr.dtype == np.int32 or arr.dtype == np.int64:
+            arr = np.asarray(arr, dtype=np.float64)
```

Comment:
The standard way to write this is `xp.isdtype(arr.dtype, 'real floating')`. What about `float16` though, was that excluded here deliberately?

Reply:
Yeah, my comment about 'the standard way' was irrelevant since we don't have array API support yet, and I failed to look at the line below for the broader floating check. Apologies for the noise!

Reply:
> > These will never be true
> 
> ```python
> x = np.asarray([1.])
> x.dtype == np.float64  # True
> ```

🤯